File: ciq_autotune/analyzers/scenario/guide.py
"""The #157 Guide / About payload — the type-level catalog behind ``/api/catalog``.

The Guide tab explains *your own* report in plain language: what every lever the
engine can attribute means, the ubiquitous-language pipeline (Anchor → Episode →
Lever → Pattern), the honest ways the engine deliberately stays quiet, and a
single worked example walked stage by stage.

This module is the **single authoring surface** for that page:

* The per-lever catalog is *generated* from :data:`~.levers._META` (title,
  exposure, recommendation, meaning) so it can never drift from the taxonomy the
  engine actually attributes. The new ``meaning`` copy lives in ``_META`` itself.
* The silence taxonomy is *enumerated* from
  :class:`~ciq_autotune.analyzers.classifiers.SilenceReason` (all seven members,
  ADR 0009) and the evidence tiers from
  :class:`~ciq_autotune.analyzers.classifiers.EvidenceTier`, each paired with the
  user-facing prose below. This page is a **read-only consumer** of those closed
  taxonomies — it mints nothing (#164/#167 own the enum; ADR 0009).
* The framing (what the engine is / won't do), the pipeline stage definitions,
  the exposure denominators, and the one worked example are authored copy held
  here — no PHI, derived from the source enums and hand-written teaching text,
  never from user data.

:func:`build_catalog` assembles the whole thing into the plain-dict payload the
Guide tab binds. It touches no store and no classifiers, so ``/api/catalog`` is a
cheap, DB-free GET.
"""

from __future__ import annotations

from ..classifiers import EvidenceTier, SilenceReason
from . import levers
from .levers import Exposure, Lever


# --- engine framing -------------------------------------------------------

_ENGINE = {
    "name": "Scenario engine",
    "tagline": (
        "It reads your own Control-IQ feed and names the most likely driver "
        "behind each high or low, one actionable cause at a time. It also marks "
        "what it's only inferring, or can't see at all."
    ),
    "is": [
        "A local, advisory read of your Control-IQ history. It explains your "
        "report; it does not drive your pump.",
        "One actionable cause per episode: the earliest driver, with the other "
        "co-occurring behaviors narrated as its consequences.",
        "Every claim is tagged Observed, Inferred, or Not-in-data, and any "
        "inference is hedged, never asserted as fact.",
    ],
    "wont": [
        "Won't dose for you or change a pump setting. It only suggests where to "
        "look.",
        "Won't invent what the feed can't show. Rescue carbs are invisible on "
        "Tandem Source, so anything leaning on them is Inferred at best.",
        "Won't pile on blame. Co-occurring behaviors collapse into one attributed "
        "lever, never a stack of separate flags.",
        "Won't grow its vocabulary casually. The lever set and the silence "
        "taxonomy are both closed, so adding one is a deliberate act.",
    ],
}


# --- the ubiquitous-language pipeline, taught by name ---------------------
# Stage colors are the locked worked-example palette (worked-split-synced.html):
# Anchor=--primary, Episode=--secondary, Lever=--accent, Pattern=--warn.

_PIPELINE = {
    "chain": [
        {
            "step": "Anchor",
            "color": "--primary",
            "one_liner": "A moment worth examining.",
            "body": "The engine scans your feed for anchors: a meal bolus, a "
                    "low, a correction, an unexplained rise. Each is a candidate "
                    "moment, not yet a judgment.",
        },
        {
            "step": "Episode",
            "color": "--secondary",
            "one_liner": "The story around an anchor.",
            "body": "Around each anchor the engine assembles an episode: the "
                    "glucose curve, the insulin on board, and every behavior that "
                    "fired in the window. That's the full context of what happened.",
        },
        {
            "step": "Lever",
            "color": "--accent",
            "one_liner": "The one actionable cause.",
            "body": "Exactly one lever is attributed per episode: its earliest "
                    "actionable driver (root-cause-by-time). Other behaviors that "
                    "also tripped are narrated as consequences of that lever, not "
                    "as separate causes.",
        },
        {
            "step": "Pattern",
            "color": "--warn",
            "one_liner": "The same lever, again and again.",
            "body": "When one lever recurs across episodes, it becomes a pattern "
                    "with a rate (\"ran away high in ~28% of your meals\"), scored "
                    "against how many opportunities of that kind you actually had.",
        },
    ],
    "suppression_truths": [
        {
            "title": "One episode, one lever.",
            "body": "A single dinner can trip three detectors at once. The engine "
                    "does not report three problems. It attributes the earliest "
                    "actionable driver and folds the rest in as its downstream "
                    "consequences. A behavior that genuinely happened can be "
                    "outranked by an earlier cause and suppressed. Fewer, truer "
                    "flags beat a longer list.",
        },
        {
            "title": "Silence is a verdict, not a gap.",
            "body": "Sometimes a behavior doesn't clear its bar. Sometimes an "
                    "observable low or suspend upstream already explains the move, "
                    "or a prior bolus owns the rise. In each case the engine stays "
                    "silent and records why. Every quiet episode carries a reason "
                    "(below). It would rather say nothing than assert something the "
                    "feed can't support.",
        },
    ],
}


# --- exposure denominators (the "how it's measured" note, shown lightly) --

_EXPOSURE_META = {
    Exposure.MEALS: (
        "Meals",
        "Scored against every meal in the window: how often a meal ran into this.",
    ),
    Exposure.LOWS: (
        "Lows",
        "Scored against every low in the window: how often a low ran into this.",
    ),
    Exposure.CORRECTION_CLUSTERS: (
        "Correction clusters",
        "Scored against every run of back-to-back corrections.",
    ),
    Exposure.HIGHS: (
        "Unexplained highs",
        "Scored against every unexplained high in the window.",
    ),
}


# --- the silence taxonomy, enumerated (all seven, ADR 0009) ---------------
# label + user-facing body per SilenceReason. The evidence tier each member
# rests on is read straight off the (member, tier) pairing below so the chip
# color matches the honesty tier the classifier already returns.

_SILENCE_META = {
    SilenceReason.INSUFFICIENT_DATA: (
        "Insufficient data",
        EvidenceTier.NOT_IN_DATA,
        "The window was too sparse to judge, with not enough readings to say "
        "anything honest.",
    ),
    SilenceReason.NO_TRIGGER: (
        "No trigger",
        EvidenceTier.OBSERVED,
        "The behavior plainly didn't happen: a flat slope, no stack, no "
        "correction, no low. A genuinely clean opportunity.",
    ),
    SilenceReason.UNDER_THRESHOLD: (
        "Under threshold",
        EvidenceTier.OBSERVED,
        "It happened but fell short of the firing bar: the near-miss where a "
        "mis-tuned threshold hides.",
    ),
    SilenceReason.UPSTREAM_CAUSE: (
        "Upstream cause",
        EvidenceTier.INFERRED,
        "An observable recent low and/or a defensive suspend explains the move. A "
        "recovery, not the behavior itself.",
    ),
    SilenceReason.PRIOR_HIGH_BASELINE: (
        "Prior high baseline",
        EvidenceTier.OBSERVED,
        "The rise started from an already-high level, not from flat, so it isn't "
        "the from-flat run this detector looks for.",
    ),
    SilenceReason.OWNED_BY_PRIOR_BOLUS: (
        "Owned by prior bolus",
        EvidenceTier.INFERRED,
        "A completed carb bolus in the recent lookback already owns the rise this "
        "dose landed into. A real excursion already covered, not a late meal "
        "bolus.",
    ),
    SilenceReason.HORIZON_EXPIRED: (
        "Horizon expired",
        EvidenceTier.OBSERVED,
        "The outcome never arrived inside the detector's window. The story ran "
        "past the edge of what it watches.",
    ),
}


# --- evidence tiers -------------------------------------------------------

_TIER_META = {
    EvidenceTier.OBSERVED: (
        "Observed",
        "A hard fact straight from the feed, like a bolus of 10 U at 15:48, or BG "
        "hitting 64. Asserted plainly.",
    ),
    EvidenceTier.INFERRED: (
        "Inferred",
        "Shape-derived, and therefore hedged (\"likely rescue carbs\"). Never "
        "asserted as fact.",
    ),
    EvidenceTier.NOT_IN_DATA: (
        "Not in data",
        "An explicit unknowable, flagged as such rather than guessed.",
    ),
}


# --- the one worked example, walked stage by stage ------------------------
# A fixed teaching episode (not user data — no PHI), authored to illustrate the
# Anchor → Episode → Lever → Pattern chain end to end. The locked layout
# (worked-split-synced.html) ties each step to its stage definition in _PIPELINE.

_WORKED = {
    "intro": "Here's how those four stages play out on one made-up dinner. It's a "
             "walkthrough to read, not feedback on your own data.",
    "episode": {
        "title": "Dinner ran away high",
        "attributed": "Carb undercount",
        "when": "Tue · 18:40 bolus · peak +2h05m",
        "peak": 247,
        "peak_state": "high",
        "spark_meta": "Bolus at **18:40**, then a **slow, sustained climb**. The "
                      "curve keeps rising long after the dose should have caught it.",
    },
    "steps": [
        {
            "stage": "Anchor",
            "t": "18:40 · meal bolus",
            "text": "You bolused **9.0 U** for dinner. A meal bolus is a candidate "
                    "moment, so the engine flags it to watch, no judgment yet.",
            "tier": "observed",
            "hedged": False,
            "cited_label": "from the feed:",
            "pills": [{"v": "9.0 U", "kind": "insulin"}, {"v": "BG 132", "kind": "cgm"}],
        },
        {
            "stage": "Episode",
            "t": "18:40 → 20:45 · the window",
            "text": "The engine assembles the context: the CGM curve, insulin "
                    "still on board, and every behavior that fired. Glucose climbs "
                    "from **132 → 247** and never turns. No correction, no low.",
            "tier": "observed",
            "hedged": False,
            "cited_label": "from the feed:",
            "pills": [{"v": "+115 mg/dL", "kind": "cgm"}, {"v": "IOB 6.1 U", "kind": "basal"}],
        },
        {
            "stage": "Lever",
            "t": "the earliest actionable driver",
            "text": "The dose landed on time and glucose was flat at the meal, yet "
                    "it still ran away. The likeliest driver is that the dose "
                    "covered **fewer carbs than the meal held**.",
            "tier": "inferred",
            "hedged": True,
            "cited_label": "reasoned from shape:",
            "pills": [{"v": "on-time bolus"}, {"v": "flat start"}, {"v": "sustained rise"}],
            "note": "A late bolus and a missed meal both also leave a rise, but "
                    "the bolus was on time and sits right on the climb, so those "
                    "are ruled out. One driver survives.",
        },
    ],
    "verdict": {
        "stage": "Lever",
        "label": "Attributed lever",
        "title": "Carb undercount",
        "body": "Exactly one lever per episode. Anything else that co-occurred is "
                "folded in as a consequence, never reported as a second problem.",
    },
    "pattern": {
        "stage": "Pattern",
        "rate": "~28%",
        "body": "This same lever recurred in ~28% of your meals this window. That "
                "rate is the pattern, scored against how many meals you actually had.",
    },
}


def _exposure_entry(exp: Exposure) -> dict:
    label, denominator = _EXPOSURE_META[exp]
    return {"value": exp.value, "label": label, "denominator": denominator}


def _lever_entry(lever: Lever) -> dict:
    exp = levers.exposure(lever)
    return {
        "value": lever.value,
        "title": levers.title(lever),
        "exposure": exp.value,
        "meaning": levers.meaning(lever),
        "recommendation": levers.recommendation(lever),
        "measured": _EXPOSURE_META[exp][1],
    }


def build_catalog() -> dict:
    """Assemble the ``/api/catalog`` payload — the type-level Guide content.

    Generated from :data:`~.levers._META` (positive catalog) and the closed
    :class:`SilenceReason` / :class:`EvidenceTier` taxonomies (ADR 0009), plus the
    authored framing/pipeline/worked-example copy above. No store access, no PHI.
    """
    return {
        "engine": _ENGINE,
        "pipeline": _PIPELINE,
        "exposures": [_exposure_entry(e) for e in Exposure],
        "levers": [_lever_entry(l) for l in Lever],
        "silence_reasons": [
            {"value": r.value, "label": _SILENCE_META[r][0],
             "tier": _SILENCE_META[r][1].value, "body": _SILENCE_META[r][2]}
            for r in SilenceReason
        ],
        "tiers": [
            {"value": t.value, "label": _TIER_META[t][0], "body": _TIER_META[t][1]}
            for t in EvidenceTier
        ],
        "worked": _WORKED,
    }

File: frontend/guide.js
/* =========================================================================
   #157 GUIDE TAB — pure render logic, extracted from the locked mockups
   (the archived catalog-cards and worked-example mocks, #157); only
   wrapped in ES `export`s here.

   VUE-FREE and DOM-FREE at import time so `node --test` can import it with no
   importmap and no DOM. The Vue component (GuideTab) lives in index.html and
   imports these helpers. The catalog payload is /api/catalog (build_catalog in
   ciq_autotune/analyzers/scenario/guide.py) — this module never hand-codes the
   lever/silence/tier copy; it only reshapes the payload for render.
   ========================================================================= */

/* ---- catalog / reference (catalog-cards-r2) ---------------------------- */

/**
 * Group the catalog's levers under their exposure denominator, in catalog
 * order, dropping any exposure with no levers. Mirrors the r2 mockup's
 * `groups()` computed — the catalog is presented grouped by "how it's
 * measured" (Meals / Lows / …), with the denominator shown once per group.
 *
 * @param {{ exposures: Array, levers: Array }} catalog
 * @returns {Array<{ value, label, denominator, levers: Array }>}
 */
export function guideGroups(catalog) {
  if (!catalog) return [];
  return catalog.exposures
    .map((e) => ({ ...e, levers: catalog.levers.filter((l) => l.exposure === e.value) }))
    .filter((g) => g.levers.length);
}

/**
 * The label for an evidence tier `value`, looked up in the payload's `tiers`
 * (so the chip text never drifts from EvidenceTier). Falls back to the raw
 * value when the tier is unknown.
 *
 * @param {Array<{ value, label }>} tiers
 * @param {string} value
 */
export function guideTierLabel(tiers, value) {
  const hit = (tiers || []).find((x) => x.value === value);
  return (hit && hit.label) || value;
}

/**
 * value -> title map from the catalog's levers. The payload owns the lever
 * titles now (retires scenario-chart.js's SCN_LEVER_TITLE): the Patterns tab's
 * scnHumanizeLever reads this map so episode-only levers get their real title.
 *
 * @param {{ levers: Array }|null} catalog
 * @returns {Object<string,string>}
 */
export function guideLeverTitles(catalog) {
  if (!catalog || !catalog.levers) return {};
  return Object.fromEntries(catalog.levers.map((l) => [l.value, l.title]));
}

/* ---- worked example (worked-split-synced) ------------------------------ */

/**
 * Render `**bold**` spans as <b>. The only markup the authored copy uses.
 * @param {string} [text]
 */
export function guideMd(text) {
  return (text || '').replace(/\*\*(.+?)\*\*/g, '<b>$1</b>');
}

/**
 * The stage definition (one_liner/body/color) for a pipeline stage, looked up
 * in the payload's pipeline chain. Falls back to a muted, empty def so the
 * worked layout never throws on an unknown stage.
 *
 * @param {Array<{ step, color, one_liner, body }>} chain
 * @param {string} stage
 */
export function guideDefFor(chain, stage) {
  return (chain || []).find((d) => d.step === stage)
    || { color: '--muted', one_liner: '', body: '' };
}

/**
 * Build the worked example's row model: each step is one row carrying its
 * stage; the verdict rides INTO the row whose stage it resolves (the Lever
 * row); the pattern is appended as its own terminal row. Mirrors the mockup's
 * `rows` computed.
 *
 * @param {{ steps: Array, verdict: object, pattern: object }} worked
 * @returns {Array<{ kind, stage, step?, verdict?, pattern? }>}
 */
export function guideWorkedRows(worked) {
  if (!worked) return [];
  const out = worked.steps.map((step) => ({
    kind: 'step',
    stage: step.stage,
    step,
    verdict: worked.verdict && worked.verdict.stage === step.stage ? worked.verdict : null,
  }));
  out.push({ kind: 'pattern', stage: worked.pattern.stage, pattern: worked.pattern });
  return out;
}

/**
 * The tie-band CSS class flags for a worked row at index `i` of `total`.
 * `first` opens the spine, `terminal` marks a verdict/pattern resolution,
 * `last-terminal` hides the trailing spine on the final row.
 *
 * @param {{ kind, verdict }} row
 * @param {number} i
 * @param {number} total
 */
export function guideRowClass(row, i, total) {
  return {
    first: i === 0,
    terminal: row.kind === 'pattern' || !!row.verdict,
    'last-terminal': i === total - 1,
  };
}

/**
 * Compute the episode sparkline geometry (polyline path, threshold band, end
 * dot) from a series of glucose points. Pure geometry, ported from the mockup's
 * IIFE so it is node-testable; the default series is the teaching curve.
 *
 * @param {number[]} [pts]
 * @param {{ w?: number, h?: number, pad?: number, band?: number }} [opts]
 */
export function guideSpark(pts = [8, 10, 12, 11, 14, 20, 30, 42, 56, 72, 86, 100, 116, 130], opts = {}) {
  const { w = 200, h = 46, pad = 4, band = 0.72 } = opts;
  const max = Math.max(...pts), min = Math.min(...pts);
  const span = max - min || 1;
  const xy = pts.map((v, i) => {
    const x = pad + (i / (pts.length - 1)) * (w - pad * 2);
    const y = h - pad - ((v - min) / span) * (h - pad * 2);
    return [x, y];
  });
  const path = xy.map((p) => p.join(',')).join(' ');
  const bandY = h - pad - (h - pad * 2) * band;
  const last = xy[xy.length - 1];
  return { w, h, path, band: `${pad},${bandY} ${w - pad},${bandY}`, dot: { x: last[0], y: last[1] } };
}
